### PR TITLE
Eliminate All Batch Calls to Deleting States

### LINKED ISSUE #4619: Explanation for High CPU / Mem Usage
Hi all,

Some users have been asking why their Prysm nodes take up so much memory and cpu. In short, there are many **extremely** expensive data copying operations we perform on the beacon chain state over 56 times per slot (a period of 12 seconds), which eats up memory at an alarming rate. Additionally, the memory footprint of keeping around eth1 deposits from the deposit contract take up a big portion of the heap profile. Thanks to @prestonvanloon for pointing this out in #4604.

In terms of CPU specifically, the biggest culprit is the function `processSlots` found in https://github.com/prysmaticlabs/prysm/blob/b030771174750c3dfad30548800f755a3bfd9ad7/beacon-chain/core/state/transition.go#L237. This function calls `stateutil.HashTreeRootState` which eats up the vast majority of CPU cycles. Not only is it slow, but it also calls `proto.Clone()` on the beacon state, often times grinding sync to a halt. Aside from this, fork choice is also a big culprit on current CPU usage.

The issues with memory usage will be fully mitigated by #4614 and fork choice will be eliminated from our bottlenecks with #4615 and other fixes @terencechain is working on. We recommend users to just hang on for a bit while we work hard on resolving these problems.

Screen Shot 2020-01-22 at 9 44 24 AM

### PR DESCRIPTION
**Pull request by rauljordan**
This resolves #4619 

---

# Description

**Write why you are making the changes in this pull request**

Even though we had a lot of memory improvements in the beacon node, certain users were still reporting massive mem consumption, which was proven by a mem flame graph extracted by one our community members on discord: 

heap

Most of the time was being spent calling bolt.Batch which only a few functions used, namely `DeleteStates`, which would likely cause a huge amount of resource consumption.

**Write a summary of the changes you are making**

This PR changes the bolt batch call to delete states to a db.Update instead to do in a single transaction. This PR also eliminates various other unused DB functions that we do not see using for the foreseeable future (_You ain't gonna need it_).

### CODE REVIEW

File: beacon-chain/db/kv/blocks.go [hunk 1]
```diff
@@ -202,58 +202,6 @@ func (k *Store) HasBlock(ctx context.Context, blockRoot [32]byte) bool {
...
-		if err := deleteValueForIndices(indicesByBucket, blockRoot[:], tx); err != nil {
-			return errors.Wrap(err, "could not delete root for DB indices")
-		}
-		k.blockCache.Del(string(blockRoot[:]))
-		return bkt.Delete(blockRoot[:])
-	})
-}
-
-// DeleteBlocks by block roots.
-func (k *Store) DeleteBlocks(ctx context.Context, blockRoots [][32]byte) error {
```
Comment (0xKiwi) [inline beacon-chain/db/kv/blocks.go]: Why was this removed?

Comment (rauljordan) [reply beacon-chain/db/kv/blocks.go]: Unused, see PR desc

**Comment by rauljordan** [comment]
Closing in favor of #4763
